lib/Controller/SocialApiController.php

```diff
@@ -0,0 +1,276 @@
+					if (in_array('basename', $social['cleanups'])) {
+						$profileId = basename($profileId);
+					}
+					// checks
+					if (in_array('number', $social['checks'])) {
+						if (!ctype_digit($profileId)) {
+							break;
+						}
+					}
+					$connector = str_replace("{socialId}", $profileId, $social['recipe']);
+					break;
+				}
+			}
+			if ($connector) {
+				break;
+			}
+		}
+		return ($connector);
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * Retrieves social profile data for a contact
+	 *
+	 * @param {String} addressbookId the addressbook identifier
+	 * @param {String} contactId the contact identifier
+	 * @param {String} type the kind of information to retrieve -- provision
+	 *
+	 * @returns {JSONResponse} an empty JSONResponse with respective http status code
+	 */
+	public function fetch(string $addressbookId, string $contactId, string $type) : JSONResponse {
+
+		$url = null;
+
+		try {
+			// get corresponding addressbook
+			$addressBooks = $this->manager->getUserAddressBooks();
+			$addressBook = null;
+			foreach($addressBooks as $ab) {
+				if ($ab->getUri() === $addressbookId) {
+					$addressBook = $ab;
+				}
+			}
+			if (is_null($addressBook)) {
+				return new JSONResponse([], Http::STATUS_INTERNAL_SERVER_ERROR);
+			}
+			
+			// search contact in that addressbook
+			$contact = $addressBook->search($contactId, ['UID'], [])[0];
+			if (is_null($contact)) {
+				return new JSONResponse([], Http::STATUS_INTERNAL_SERVER_ERROR); 
+			}
+
+			// get social data
+			$socialprofile = $contact['X-SOCIALPROFILE'];
+			if (is_null($socialprofile)) {
+				return new JSONResponse([], Http::STATUS_INTERNAL_SERVER_ERROR);
+			}
+
+			// retrieve data
+			try {
+				$url = $this->getSocialConnector($socialprofile);
+			}
+			catch (Exception $e) {
+				return new JSONResponse([], Http::STATUS_BAD_REQUEST);
+			}
+
+			if (empty($url)) {
+				return new JSONResponse([], Http::STATUS_NOT_IMPLEMENTED);
+			}
+
+			$host = parse_url($url);
+			if (!$host) {
+				return new JSONResponse([], Http::STATUS_BAD_REQUEST);
+			}
+			$opts = [
+				"http" => [
+					"method" => "GET",
+					"header" => "User-Agent: Nextcloud Contacts App"
+				]
+			];
+			$context = stream_context_create($opts);
+			$socialdata = file_get_contents($url, false, $context);
+
+			$image_type = null;
+			foreach ($http_response_header as $value) {
+				if (preg_match('/^Content-Type:/i', $value)) {
+					if (stripos($value, "image") !== false) {
+						$image_type = substr($value, stripos($value, "image"));
+					}
+				}
+			}
+
+			if ((!$socialdata) || ($image_type === null)) {
+				return new JSONResponse([], Http::STATUS_NOT_FOUND);
+			}
+
+			// update contact
+			switch ($type) {
+				case 'avatar':
+					if (!empty($contact['PHOTO'])) {
+						// overwriting without notice!
+					}
+					$changes = array();
+					$changes['URI']=$contact['URI'];
+					$changes['PHOTO'] = "data:" . $image_type . ";base64," . base64_encode($socialdata);
```

Comment:
This syntax is not compatible with all vcard versions :)

Reply:
ahw, would have been too easy... but I suppose there is a function that I can use? Will look into that tonight.

Reply:
No idea :)
We only support vcard 3 & 4 for now, so it's only two different syntaxes ;)
If the version is not one of them, you can throw (should not happen, but you're never too careful :stuck_out_tongue_closed_eyes: 

https://tools.ietf.org/html/rfc6350#section-6.2.4
https://tools.ietf.org/html/rfc2426#section-3.1.4